res/controllers/Pioneer-DDJ-200-scripts.js

```diff
@@ -45,123 +45,366 @@ Deck n = 0/1:                                   Midi in         Midi out
+    indicatorStyle: "indicator_250ms",
+});
+
+DDJ200.DoubleRingBuffer = class {
+    constructor(indexableUnshifted, indexableShifted) {
+        this.indexableUnshifted = indexableUnshifted;
+        this.indexableShifted = indexableShifted;
+        this.indexable = this.indexableUnshifted;
+        this.index = 0;
+        this.isShifted = false;
+    };
+    shift() {
+        this.indexable = this.indexableShifted;
+        this.index = 0;
+        this.isShifted = true;
+    };
+    unshift() {
+        this.indexable = this.indexableUnshifted;
+        this.index = 0;
+        this.isShifted = false;
+    };
+    current() {
+        return this.indexable[this.index];
+    };
+    next() {
+        this.index = script.posMod(this.index + 1, this.indexable.length);
+        return this.current();
+    };
+};
+
+DDJ200.PadModeContainers.ModeSelector = function(group) {
+    const deckOffset = (((script.deckFromGroup(group) - 1) % 2) * 2);
+
+    const updateSelectorLeds = () => {
+        this.forEachComponent(c => c.trigger());
+    };
+
+    const setPads = newPadInstance => {
+        if (newPadInstance === this.choosenPadInstance) {
+            return;
+        }
+
+        this.choosenPadInstance.forEachComponent(function(component) {
+            component.disconnect();
+        });
+        this.choosenPadInstance = newPadInstance;
+        updateSelectorLeds();
+        this.choosenPadInstance.reconnectComponents(function(c) {
+            if (c.group === undefined) {
+                c.group = group;
+            }
+        });
+    };
+
+    this.input = () => {
+        if (this.padInstancesBuffer.isShifted) {
+            this.padInstancesBuffer.unshift();
+            setPads(this.padInstancesBuffer.current());
+        } else {
+            setPads(this.padInstancesBuffer.next());
+        }
+    };
+
+    this.shiftedInput = () => {
+        if (!this.padInstancesBuffer.isShifted) {
+            this.padInstancesBuffer.shift();
+            setPads(this.padInstancesBuffer.current());
+        } else {
+            setPads(this.padInstancesBuffer.next());
+        }
+    };
+
+    this.startupModeInstance = new DDJ200.PadModeContainers.Hotcue(deckOffset);
+    this.loopInstance = new DDJ200.PadModeContainers.Loop(deckOffset);
+    this.effectInstance = new DDJ200.PadModeContainers.Effect(deckOffset);
+    this.jumpInstance = new DDJ200.PadModeContainers.Jump(deckOffset);
+
+    this.padInstancesBuffer = new DDJ200.DoubleRingBuffer(
+        [this.startupModeInstance, this.loopInstance, this.effectInstance],
+        [this.jumpInstance]
+    );
+
+    this.choosenPadInstance = new DDJ200.PadMode();
+    setPads(this.startupModeInstance);
+};
+DDJ200.PadModeContainers.ModeSelector.prototype = new components.ComponentContainer();
+
+DDJ200.init = function() {
+    this.Decks = {
+        left: new DDJ200.Deck([1, 3], 1),
+        right: new DDJ200.Deck([2, 4], 2),
+    };
```

Comment:
NIT:
`TitleCase` is usually reserved for class names, use `sname_case` instead. Also if you make this a `ComponentContainer` you can replace all the `Object.values(DDJ200.Decks).forEach(deck => ...);` calls with a slightly nicer `DDJ200.Decks.forEachComponent(deck => ...);`


```suggestion
    this.decks = new components.ComponentContainer({
        left: new DDJ200.Deck([1, 3], 1),
        right: new DDJ200.Deck([2, 4], 2),
    });
```

Reply:
fixed with 42f5342be24c467449df6092cd1b9117b5ef5381